### PR TITLE
Feature/pre train subreddit

### PR DESCRIPTION
**Pull request by Jaxing**
This enables to pre-train the network on different output labels for the same input data. Specifically to use subreddits as lables for pre-training.
In main.py the a new layer is added after adding secondary output layer, this is so that the network can generelise from the learnt prameters and when tested showed better preformance. Where in the network the secondary output should be placed should be treated as a hyperparameter.

### CODE REVIEW

File: model/model.py [hunk 1]
```diff
@@ -184,45 +189,46 @@ def validate_batch(self):
 
     # TODO funktionen gör alldeles för mycket,
     # dela upp utskrift, beräkning och träning
-    def train(self):
+    def train(self, use_pretrained_net=False):
         """ Trains the model on the dataset """
         print("Starting training...")
 
-        if self.use_pretrained:
+        if self.use_pretrained and \
```
Comment (hsson) [inline model/model.py]: This if statement can be simplified to:
```
if self.use_pretrained and (self.use_pretrained_net == use_pretrained_net)
```

File: config.template.yaml [hunk 2]
```diff
@@ -11,6 +11,7 @@ network:
     validation_data: 'validation_data_top_n_single.csv'
     training_data: 'training_data_top_n_single.csv'
     testing_data: 'testing_data_top_n_single.csv'
+    pre_train_data: 'pre-train_data_top_n.csv'
```
Comment (hsson) [inline config.template.yaml]: Remove

File: main.py [hunk 3]
```diff
@@ -38,7 +39,22 @@ def main():
         config_file = networkconfig[conf]
         with tf.Session() as sess:
             builder = ModelBuilder(config_file, sess)
+            builder.add_input_layer()
+
+            # Add a number of hidden layers
```
Comment (hsson) [inline main.py]: Why is this done in main.py? Shouldn't it be done in the build function of the model builder?

File: main.py [hunk 4]
```diff
@@ -38,7 +39,22 @@ def main():
...
+
+            if builder._model.use_pretrained_net:
+                builder.add_secondary_output()
+                builder.add_layer(config_file[HIDDEN_NEURONS])
+
+            builder.add_output_layer()\
+                .add_precision_operations()
+
             network_model = builder.build()
+            if config_file[USE_PRETRAINED_NET]:
```
Comment (hsson) [inline main.py]: Can't this be moved to the model itself? To the beginning of the train function maybe? Just tryting to keep the main.py file as clean as possible :smile:

File: model/model.py (hunk 1 above)
Comment (hsson) [reply model/model.py]: Also, what does this actually mean? This pull request doesn't change behavior of the embedding matrix right? So why is it required that  self.use_pretrained_net == use_pretrained_net ? And why are there two separate variables for use_pretrained_net?

File: model/model.py [hunk 5]
```diff
@@ -184,45 +189,46 @@ def validate_batch(self):
...
+                (self.use_pretrained_net and use_pretrained_net) or \
+                (not self.use_pretrained_net and not use_pretrained_net):
             self._session.run(self.embedding_init,
                               feed_dict={self.embedding_placeholder:
                                          self.data.embedding_matrix})
-        self.train_writer = \
```
Comment (hsson) [inline model/model.py:195]: Why are the writers moved to the model builder?

File: model/model_builder.py [hunk 6]
```diff
@@ -178,6 +183,49 @@ def add_output_layer(self):
 
         return self
 
+    def add_secondary_output(self):
```
Comment (hsson) [inline model/model_builder.py]: This is very similar to the other output function. Is it possible to combine them and maybe take the labels as an input or something?

File: model/util/csv_reader.py [hunk 7]
```diff
@@ -34,7 +34,7 @@ class Dataenum(Enum):
     TESTING = "testing_data"
     TRAINING = "training_data"
     VALIDATION = "validation_data"
-
+    PRE_TRAINING = "pre_training_data"
```
Comment (hsson) [inline model/util/csv_reader.py]: remove

File: model/model.py [hunk 8]
```diff
@@ -67,13 +67,17 @@ def __init__(self, config, session):
         self.use_constant_limit = config[USE_CONSTANT_LIMIT]
         self.constant_prediction_limit = config[CONSTANT_PREDICTION_LIMIT]
         self.use_concat_input = config[USE_CONCAT_INPUT]
+        self.use_pretrained_net = config[USE_PRETRAINED_NET]
+        self.subreddit_count = 0
```
Comment (hsson) [inline model/model.py:72]: Is this variable used?

File: model/model.py (hunk 1 above)
Comment (Jaxing) [reply model/model.py]: We only want to initialise the embedding matrix once but we run train twice therefore we want to run it only if both self.use_pretrained_net and use_pretrained_net is true i.e. we want to pre-train on subreddits and this call to train will pre-train or if both of them are false i.e. we don't want to pre-train and this call don't do pre-train. 
That's just bad naming, the class variable is whether the model should use pre-training on the network and the parameter is whether this call to the method should pre-train the network

**Comment by hsson** [comment]
Also, I just realised that the concatenated subreddit input is still used when pre-training on subreddits. Maybe this should not be concatenated when doing that? Seems a bit biased. At the same time... The dimensions of the weights depend the size of the input.

File: model/model.py (hunk 5 above)
Comment (Jaxing) [reply model/model.py:195]: Perhaps not necessary but otherwise the writer initiated twice, once for each time we train.

File: main.py (hunk 3 above)
Comment (Jaxing) [reply main.py]: This is more modular, if someone want to set up the network in another way than how it is done in the build method then there is no way of building the model

File: model/model.py (hunk 8 above)
Comment (Jaxing) [reply model/model.py:72]: yes

File: model/model_builder.py (hunk 6 above)
Comment (Jaxing) [reply model/model_builder.py]: Might be, one problem is that we don't want to overwrite the training op for either could maybe be dealt with using a boolean

**Comment by hsson** [comment]
- Use softmax for subreddit predictions.
- Output layers should be last

**Comment by hsson** [comment]
Feedback from meeting has been implemented and tested.
